### PR TITLE
Master merge

### PR DESCRIPTION
**Pull request by mgloystein**

### CODE REVIEW

File: adapters/beachfront/beachfront.go [hunk 1]
```diff
@@ -175,6 +176,15 @@ func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapte
 	headers.Add("Content-Type", "application/json;charset=utf-8")
 	headers.Add("Accept", "application/json")
 
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
```
Comment (bluedreamer) [inline adapters/beachfront/beachfront.go:181]: Don't know if it matters but x-forwarded-for can be a comma separated list of IP's

File: adapters/beachfront/beachfront.go [hunk 2]
```diff
@@ -175,6 +176,15 @@ func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapte
...
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
```
Comment (bluedreamer) [inline adapters/beachfront/beachfront.go:184]: Hmm we also have an opt_out cookies and a gdpr opt out cookie. Which also begs the question of "Do we have the spotx "audience" cookie at this point

File: adapters/beachfront/beachfront.go [hunk 3]
```diff
@@ -394,12 +401,33 @@ func getVideoRequest(req *openrtb.BidRequest) (BeachfrontVideoRequest, []error,
 func (a *BeachfrontAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
 	var bids []openrtb.Bid
 	var bidtype = getBidType(internalRequest)
+
+	/*
+		Beachfront is now sending an empty array and 200 as their "no results" response. This should catch that.
```
Comment (bluedreamer) [inline adapters/beachfront/beachfront.go:406]: Usually openrtb return "204 no content"

File: adapters/beachfront/beachfront.go [hunk 4]
```diff
@@ -499,6 +527,13 @@ func extractVideoCrid(nurl string) string {
 	return strings.TrimSuffix(chunky[2], ":")
 }
 
+// Thank you, brightroll.
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
```
Comment (bluedreamer) [inline adapters/beachfront/beachfront.go:533]: Should this code test if the header already exists? eg you can only have 1 Location: header

File: adapters/improvedigital/improvedigital.go [hunk 5]
```diff
@@ -58,15 +59,36 @@ func (a *ImprovedigitalAdapter) MakeBids(internalRequest *openrtb.BidRequest, ex
...
+
+	if len(bidResp.SeatBid) > 1 {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected SeatBid! Must be only one but have: %d", len(bidResp.SeatBid)),
+		}}
+	}
+
+	seatBid := bidResp.SeatBid[0]
+
+	if len(seatBid.Bid) == 0 {
```
Comment (bluedreamer) [inline adapters/improvedigital/improvedigital.go:74]: Not understanding go very well, but isn't the check at 74 esentially the same as 62? Strings are slices right?

File: stored_requests/multifetcher.go [hunk 6]
```diff
@@ -36,10 +36,10 @@ func (mf MultiFetcher) FetchRequests(ctx context.Context, requestIDs []string, i
 	return
 }
 
-func (mf MultiFetcher) FetchCategories(primaryAdServer, publisherId, iabCategory string) (string, error) {
+func (mf MultiFetcher) FetchCategories(ctx context.Context, primaryAdServer, publisherId, iabCategory string) (string, error) {
 	for _, f := range mf {
 		if cf, ok := f.(CategoryFetcher); ok {
-			iabCategory, _ := cf.FetchCategories(primaryAdServer, publisherId, iabCategory)
+			iabCategory, _ := cf.FetchCategories(ctx, primaryAdServer, publisherId, iabCategory)
```
Comment (bluedreamer) [inline stored_requests/multifetcher.go:42]: I don't know if this matters, but we do not generally use iab categories internally. so not sure of the source of this data. We do though have an endpoint that will map them to IAB ones

**Comment by mgloystein** [comment]
This is purely a merge from the root master project, there are not spotx changes in this PR.

**Review by bluedreamer — APPROVED** [review]

**Pull request merged**